File: admin/apps/flutter/lib/app/pages/member/member_controller.dart
/*
 * Copyright (c) 2026 erik <erik@erik.xyz> — https://erik.xyz
 */

import 'package:get/get.dart';
import '../../services/api_service.dart';

/// 会员管理控制器（只读：列表 + 详情）
/// 契约对应 admin/app/admin/controller/MemberController.php：
///   GET /admin/members?page&limit&nickname&phone&uid&member_level&reg_date_start&reg_date_end
///   GET /admin/members/{id} → data.user + data.orders{list,total,page,limit} + data.active_cards[]
/// 后端未注册新增/编辑路由（resource 仅 index/show），此页只读。
/// 列表项：id(hashid), nickname, phone(脱敏), real_name(脱敏), member_level,
///         total_spent, order_count, member_cards_count, created_at
/// 详情 active_cards 项：id, user_id, card_id, start_at, end_at,
///         total_times, used_times, status, member_card{name,...}
class MemberController extends GetxController {
  final api = ApiService();

  final members = <dynamic>[].obs;
  final isLoading = false.obs;
  final total = 0.obs;
  final page = 1.obs;
  final limit = 15.obs;
  final keyword = ''.obs;

  @override
  void onInit() {
    super.onInit();
    loadMembers();
  }

  Future<void> loadMembers({bool reset = false}) async {
    if (reset) page.value = 1;
    isLoading.value = true;
    try {
      final params = <String, dynamic>{
        'page': page.value,
        'limit': limit.value,
      };
      if (keyword.value.isNotEmpty) params['nickname'] = keyword.value;

      final resp = await api.get('/admin/members', params: params);
      members.value = resp['data']['list'] as List<dynamic>;
      total.value = resp['data']['total'] as int;
    } catch (e) {
      Get.snackbar('错误', '加载会员列表失败: $e');
    } finally {
      isLoading.value = false;
    }
  }

  Future<void> search(String kw) async {
    keyword.value = kw.trim();
    await loadMembers(reset: true);
  }

  Future<void> nextPage() async {
    if (page.value * limit.value < total.value) {
      page.value++;
      await loadMembers();
    }
  }

  Future<void> prevPage() async {
    if (page.value > 1) {
      page.value--;
      await loadMembers();
    }
  }

  /// 会员详情（用户信息 + 有效会员卡 + 订单历史）
  Future<Map<String, dynamic>?> loadDetail(String id) async {
    try {
      final resp = await api.get('/admin/members/$id');
      return resp['data'] as Map<String, dynamic>;
    } catch (e) {
      Get.snackbar('错误', '加载会员详情失败: $e');
      return null;
    }
  }

  static String cardStatusLabel(dynamic status) {
    final s = status?.toString();
    if (s == 'active') return '有效';
    if (s == 'expired') return '已过期';
    if (s == 'used_up') return '次数用完';
    return s ?? '-';
  }
}

File: admin/app/admin/controller/MemberController.php
<?php
/*
 * Copyright (c) 2026 erik <erik@erik.xyz> — https://erik.xyz
 */

declare(strict_types=1);

namespace app\admin\controller;

use app\model\User;
use app\model\Order;
use app\model\UserMemberCard;
use support\Request;
use support\Response;

class MemberController extends BaseController
{
    /**
     * 会员列表（有订单的用户）
     * 搜索: uid / phone / nickname / reg_date
     */
    public function index(Request $request): Response
    {
        $page  = (int) $request->input('page', 1);
        $limit = (int) $request->input('limit', 15);
        $uid   = $request->input('uid', '');
        $phone = $request->input('phone', '');
        $nickname = $request->input('nickname', '');
        $regDateStart = $request->input('reg_date_start', '');
        $regDateEnd   = $request->input('reg_date_end', '');
        $memberLevel = $request->input('member_level');

        $query = User::whereHas('orders');

        if ($uid) {
            $query->where('id', $uid);
        }
        if ($phone) {
            $query->where('phone', 'like', "%{$phone}%");
        }
        if ($nickname) {
            $query->where('nickname', 'like', "%{$nickname}%");
        }
        if ($regDateStart) {
            $query->whereDate('created_at', '>=', $regDateStart);
        }
        if ($regDateEnd) {
            $query->whereDate('created_at', '<=', $regDateEnd);
        }
        if ($memberLevel !== null && $memberLevel !== '') {
            $query->where('member_level', (string) $memberLevel);
        }

        $total = $query->count();
        $list  = $query->offset(($page - 1) * $limit)
                       ->limit($limit)
                       ->orderBy('id', 'desc')
                       ->get()
                       ->map(function ($user) {
                           $data = $user->toArray();
                           // 脱敏
                           if (!empty($data['phone'])) {
                               $data['phone'] = preg_replace('/^(\d{3})\d+(\d{4})$/', '$1****$2', $data['phone']);
                           }
                           if (!empty($data['real_name'])) {
                               $data['real_name'] = mb_substr($data['real_name'], 0, 1) . '**';
                           }
                           // 会员统计
                           $data['total_spent'] = Order::where('user_id', $user->id)
                               ->where('status', 'completed')
                               ->sum('paid_amount');
                           $data['order_count'] = Order::where('user_id', $user->id)->count();
                           $data['member_cards_count'] = UserMemberCard::where('user_id', $user->id)
                               ->where('status', 'active')->count();

                           return $this->encodeIds($data);
                       });

        return $this->success([
            'list'  => $list,
            'total' => $total,
            'page'  => $page,
            'limit' => $limit,
        ]);
    }

    /**
     * 会员详情 + 订单历史 + 有效会员卡
     */
    public function show(Request $request, string $hashid): Response
    {
        $id   = $this->decodeId($hashid);
        $user = User::find($id);
        if (!$user) {
            return $this->fail('用户不存在', 404);
        }

        $data = $user->toArray();
        if (!empty($data['phone'])) {
            $data['phone'] = preg_replace('/^(\d{3})\d+(\d{4})$/', '$1****$2', $data['phone']);
        }
        if (!empty($data['real_name'])) {
            $data['real_name'] = mb_substr($data['real_name'], 0, 1) . '**';
        }

        // 订单历史
        $orderPage   = (int) $request->input('order_page', 1);
        $orderLimit  = (int) $request->input('order_limit', 10);
        $ordersQuery = Order::where('user_id', $id)->with(['items', 'payment']);
        $orderTotal  = $ordersQuery->count();
        $orders      = $ordersQuery->offset(($orderPage - 1) * $orderLimit)
                                   ->limit($orderLimit)
                                   ->orderBy('id', 'desc')
                                   ->get()
                                   ->map(fn($o) => $this->encodeIds($o->toArray()));

        // 有效会员卡
        $cards = UserMemberCard::where('user_id', $id)
            ->where('status', 'active')
            ->with('memberCard')
            ->get()
            ->map(fn($c) => $this->encodeIds($c->toArray()));

        // 消费统计
        $data['total_spent'] = Order::where('user_id', $id)->where('status', 'completed')->sum('paid_amount');

        return $this->success([
            'user'   => $this->encodeIds($data),
            'orders' => [
                'list'  => $orders,
                'total' => $orderTotal,
                'page'  => $orderPage,
                'limit' => $orderLimit,
            ],
            'active_cards' => $cards,
        ]);
    }

    /**
     * 设置会员等级
     */
    public function updateLevel(Request $request, string $hashid): Response
    {
        $id   = $this->decodeId($hashid);
        $user = User::find($id);
        if (!$user) {
            return $this->fail('用户不存在', 404);
        }

        $level = (int) $request->input('member_level', 0);
        $user->member_level = $level;
        $user->save();

        return $this->success($this->encodeIds($user->toArray()), '会员等级更新成功');
    }

    /**
     * 更改推荐人
     */
    public function updateParent(Request $request, string $hashid): Response
    {
        $id   = $this->decodeId($hashid);
        $user = User::find($id);
        if (!$user) {
            return $this->fail('用户不存在', 404);
        }

        $referrerId = $request->input('referrer_id', '');
        if (empty($referrerId)) {
            $user->referrer_id = null;
        } else {
            $referrer = User::find($referrerId);
            if (!$referrer) {
                return $this->fail('推荐人不存在', 404);
            }
            if ($referrerId === (string) $id) {
                return $this->fail('不能将自己设为推荐人', 422);
            }
            $user->referrer_id = $referrerId;
        }
        $user->save();

        return $this->success($this->encodeIds($user->toArray()), '推荐人更新成功');
    }
}
